src/kafunk/Utility/AsyncSeq.fs

```diff
@@ -255,4 +255,39 @@ module AsyncSeq =
 
   let mergeChoice4 (s1:AsyncSeq<'a>) (s2:AsyncSeq<'b>) (s3:AsyncSeq<'c>) (s4:AsyncSeq<'d>) : AsyncSeq<Choice<'a, 'b, 'c, 'd>> =
     AsyncSeq.mergeAll 
-      [ s1 |> AsyncSeq.map Choice1Of4 ; s2 |> AsyncSeq.map Choice2Of4 ; s3 |> AsyncSeq.map Choice3Of4 ; s4 |> AsyncSeq.map Choice4Of4 ]
\ No newline at end of file
+      [ s1 |> AsyncSeq.map Choice1Of4 ; s2 |> AsyncSeq.map Choice2Of4 ; s3 |> AsyncSeq.map Choice3Of4 ; s4 |> AsyncSeq.map Choice4Of4 ]
+
+  /// merges multiple async seqs together, distribution spread across evenly throughout each async seq
+  let mergeAllRoundRobin (ss:AsyncSeq<'T> list) : AsyncSeq<'T> =
+    asyncSeq { 
+      let n = ss.Length
+      if n > 0 then 
+        let ies = [| for source in ss -> source.GetEnumerator()  |]
+        let tasks = Array.zeroCreate n
+        for i in 0 .. ss.Length - 1 do 
+          let! task = Async.StartChildAsTask (ies.[i].MoveNext())
+          do tasks.[i] <- task
+        let fin = ref n
+        while fin.Value > 0 do 
+          let! results = 
+            tasks 
+            |> Array.map(fun t -> async { 
+              let! ti = Task.WhenAny([t]) |> Async.AwaitTask
+              let i  = Array.IndexOf (tasks, ti)
+              let v = ti.Result
+              match v.IsSome with 
+              | true -> 
+                  let! task = Async.StartChildAsTask (ies.[i].MoveNext())
+                  do tasks.[i] <- task
+              | false ->
+                  let t = System.Threading.Tasks.TaskCompletionSource()
+                  tasks.[i] <- t.Task // result never gets set
+                  fin := fin.Value - 1
+              return v
+               })
+            |> Async.Parallel         
+          for res in results do
+            if res.IsSome then
+              yield res.Value
+            
+    }
```

Comment:
How about this alternative implementation that would solve the potential hanging issue?

```
let itasks = Array.zeroCreate n
for i in 0..ss.Length - 1 do
  let! task = Async.StartChildAsTask (ies.[i].MoveNext())
  do itasks.[i] <- i,task


let fin = ref n
while fin.Value > 0 do
  let! results =
    itasks
    |> Array.map (fun (i,task) -> async {
        if i >= 0 then
          let! t = task |> Async.AwaitTask
          return (i, t.Result)
        else
          return (i, None) })
    |> Async.Parallel

  for j in 0..results.Length-1 do
    let i, res = results.[j]
    match res with
    | Some r ->
      yield r
      let! task = Async.StartChildAsTask (ies.[i].MoveNext())
      do itasks.[i] <- i,task
    | None when i >= 0 ->
      let _,unused = itasks[j]
      do itasks <- -1,unused
      fin := fin.Value - 1
    | _ -> () // ignore
```

This implementation keeps track of the indices of tasks corresponding to asyncseqs that have not finished and run them in parallel. There's a more succinct implementation where we filter out finished asyncseqs as we progress using something like `Array.filter` but i think that would have resulted in too many array copies for long asyncseqs. I believe this one should have the same overhead as the original `AsyncSeq.mergeAll`

Reply:
Is there a reason that an approach like this wouldn't work? Quite a bit simpler, and I don't think there would be any additional overhead.

```
  let mergeAllRoundRobin (asyncSeqList:AsyncSeq<'T> list) : AsyncSeq<'T> =
    asyncSeq {
        let queue = new Queue<AsyncSeq<'T>>(asyncSeqList)
        while queue.Count > 0 do
            let next = queue.Dequeue()
            let! firstValue = AsyncSeq.tryFirst next 
            if firstValue.IsSome then
                queue.Enqueue(next)
                yield firstValue.Value
    }
```

Reply:
@egracer, yeah, because `AsyncSeq.tryFirst` always gets a new enumerator (to try to get first). So you'll only get the first element of each sequence.

https://github.com/fsprojects/FSharp.Control.AsyncSeq/blob/94c2f487eb509b6f18093c0716cbb8b54c365125/src/FSharp.Control.AsyncSeq/AsyncSeq.fs#L816

Reply:
Maybe this moves us closer?

```

  let mergeAllRoundRobin (asyncSeqList:AsyncSeq<'T> list) : AsyncSeq<'T> =
    asyncSeq {
        let asyncEnumerables = asyncSeqList |> List.map (fun s -> s.GetEnumerator())
        let queue = new Queue<IAsyncEnumerator<'T>>(asyncEnumerables)
        while queue.Count > 0 do
            let next = queue.Dequeue()
            let! firstValue = next.MoveNext()
            if firstValue.IsSome then
                queue.Enqueue(next)
                yield firstValue.Value
    }
```

This would achieve a strict round-robin of the async seqs if that's what we're looking for, but also means that if we have asyncs [A; B; C;] where A takes a really long time but B and C are fast, then we might block the returned async seq for awhile each time A returns to the queue. Not sure if that's the desired behavior. If not, we could use a concurrent queue. I guess I'm leaning away from Task logic here because I find it very hard to follow.